flytepropeller/pkg/compiler/validators/utils.go

```diff
@@ -170,125 +169,244 @@ func UnionDistinctVariableMaps(m1, m2 map[string]*core.Variable) (map[string]*co
+}
+
+type trivialInstanceChecker struct {
+	literalType *core.LiteralType
+}
+
+func (t trivialInstanceChecker) isInstance(lit *core.Literal) bool {
+	if _, ok := lit.GetValue().(*core.Literal_Scalar); !ok {
+		return false
+	}
+	targetType := t.literalType
+	if targetType.GetEnumType() != nil {
+		// If t is an enum, it can be created from a string as Enums as just constrained String aliases
+		if _, ok := lit.GetScalar().GetPrimitive().GetValue().(*core.Primitive_StringValue); ok {
+			return true
 		}
 	}
-	unionLiteralType := &core.LiteralType{
-		Type: &core.LiteralType_UnionType{
-			UnionType: &core.UnionType{
-				Variants: variants,
-			},
-		},
+
+	literalType := literalTypeForScalar(lit.GetScalar())
+	err := ValidateLiteralType(literalType)
+	if err != nil {
+		return false
+	}
+	return AreTypesCastable(literalType, targetType)
+}
+
+type noneInstanceChecker struct{}
+
+func (t noneInstanceChecker) isInstance(lit *core.Literal) bool {
+	if lit == nil {
+		return true
 	}
+	_, ok := lit.GetScalar().GetValue().(*core.Scalar_NoneType)
+	return ok
+}
 
-	if isNested {
-		return &core.LiteralType{
-			Type: &core.LiteralType_CollectionType{
-				CollectionType: unionLiteralType,
-			},
+type collectionInstanceChecker struct {
+	literalType *core.LiteralType
+}
+
+func (t collectionInstanceChecker) isInstance(lit *core.Literal) bool {
+	if _, ok := lit.GetValue().(*core.Literal_Collection); !ok {
+		return false
+	}
+	for _, x := range lit.GetCollection().GetLiterals() {
+		if !IsInstance(x, t.literalType.GetCollectionType()) {
+			return false
 		}
 	}
+	return true
+}
 
-	return unionLiteralType
+type mapInstanceChecker struct {
+	literalType *core.LiteralType
 }
 
-func literalTypeForLiterals(literals []*core.Literal) *core.LiteralType {
-	innerType := make([]*core.LiteralType, 0, 1)
-	innerTypeSet := sets.NewString()
-	var noneType *core.LiteralType
-	for _, x := range literals {
-		otherType := LiteralTypeForLiteral(x)
-		otherTypeKey := otherType.String()
-		if _, ok := x.GetValue().(*core.Literal_Collection); ok {
-			if x.GetCollection().GetLiterals() == nil {
-				noneType = otherType
-				continue
-			}
+func (t mapInstanceChecker) isInstance(lit *core.Literal) bool {
+	if _, ok := lit.GetValue().(*core.Literal_Map); !ok {
+		return false
+	}
+	for _, x := range lit.GetMap().GetLiterals() {
+		if !IsInstance(x, t.literalType.GetMapValueType()) {
+			return false
 		}
+	}
+	return true
+}
+
+type blobInstanceChecker struct {
+	literalType *core.LiteralType
+}
+
+func (t blobInstanceChecker) isInstance(lit *core.Literal) bool {
+	if _, ok := lit.GetScalar().GetValue().(*core.Scalar_Blob); !ok {
+		return false
+	}
+
+	blobType := lit.GetScalar().GetBlob().GetMetadata().GetType()
+	if blobType == nil {
+		return false
+	}
+
+	// Empty blobs should match any blob.
+	if blobType.GetFormat() == "" || t.literalType.GetBlob().GetFormat() == "" {
+		return true
+	}
+
+	return blobType.GetFormat() == t.literalType.GetBlob().GetFormat()
```

Comment:
I think we should also add a check to make sure `dimensionality` also matches.  (i.e. FlyteFile should not be an instance of FlyteDirectory)